python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -1175,21 +1162,109 @@ class PyInferenceRequest
   {
     ThrowIfError(TRITONSERVER_InferenceRequestCancel(triton_object_));
   }
 
+  // Response management
+  void GetNextResponse(const py::object& py_future)
+  {
+    // Called exclusively from Python, so GIL is always held.
+    std::lock_guard lock(response_mu_);
 
- public:
-  std::unique_ptr<CallbackResource> request_callback_resource_{nullptr};
+    // There are two cases:
+    // 1. The backend is faster than the frontend:
+    //      There will be response(s) queued at 'responses_' deque, so pop the
+    //      response from the deque and set it to the 'py_future' object.
+    // 2. The backend is slower than the frontend:
+    //      The 'responses_' deque will be empty, so track the 'py_future' in
+    //      this request object until the 'AddNextResponse()' is called.
+    if (responses_.empty()) {
+      if (response_future_.ptr() != nullptr) {
+        throw AlreadyExistsError("cannot call GetNextResponse concurrently");
+      }
+      response_future_ = py_future;
+    } else {
+      std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t>&
+          py_response = responses_.front();
+      PyFutureSetResult(py_future, py_response);
+      responses_.pop_front();
+    }
+  }
 
  private:
-  std::unique_ptr<PyResponseAllocator::CallbackResource>
-      allocator_callback_resource_{nullptr};
-  std::unique_ptr<PyInferenceResponse::CallbackResource>
-      response_callback_resource_{nullptr};
+  void AddNextResponse(
+      struct TRITONSERVER_InferenceResponse* response, const uint32_t flags)
+  {
+    // Pair the 'response' and 'flags' so they can be easily moved around.
+    std::shared_ptr<PyInferenceResponse> managed_ptr(nullptr);
+    if (response != nullptr) {
+      managed_ptr.reset(new PyInferenceResponse(response, true /* owned */));
+    }
+    std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t> py_response(
+        std::move(managed_ptr), std::move(flags));
+
+    // There are two cases:
+    // 1. The backend is faster than the frontend:
+    //      The 'response_future_' will be empty since the frontend is not
+    //      waiting on a response, so add the 'py_response' to the 'responses_'
+    //      deque.
+    // 2. The backend is slower than the frontend:
+    //      The 'response_future_' will be set since the frontend is waiting on
+    //      a response, so set the 'py_response' to the awaiting
+    //      'response_future_'.
+    {
+      std::lock_guard lock(response_mu_);
+      if (response_future_.ptr() == nullptr) {
+        responses_.push_back(std::move(py_response));
+        return;
+      }
+    }
+    // There is no need to hold the 'response_mu_' after knowing the frontend is
+    // waiting, given there should be no concurrent calls into
+    // 'GetNextResponse', the frontend will wait until a result is set for the
+    // 'response_future_'.
+    {
+      py::gil_scoped_acquire gil;
+      py::object response_future_local(std::move(response_future_));
```

Comment:
still a little unsure why we need this is GetNextResponse holds the gil - but I think we can explore that seperately (the explicit move and release of the response_future_)

Reply:
the GIL is required when the `response_future_local` object goes out of scope, which decrement the ref count to the future object.